hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/om/KeyManagerImpl.java

```diff
@@ -2205,6 +2272,167 @@ private void listStatusFindKeyInTableCache(
+        OzoneFileStatus fileStatusInfo = getOzoneFileStatusV1(volumeName,
+                bucketName, startKey, false, null);
+        if (fileStatusInfo != null) {
+          prefixKeyInDB = fileStatusInfo.getKeyInfo().getParentObjectID();
+          seekKeyInDB = prefixKeyInDB + OZONE_URI_DELIMITER
+                  + fileStatusInfo.getKeyInfo().getFileName();
+        }
+      }
+
+      // Not required to search in TableCache because all the deleted keys
+      // are marked directly in directory table or in key table by breaking
+      // the pointer to its sub-dirs. So, there is no issue of inconsistency.
+      int countEntries = 0;
+      // Seek the given key in key table.
+      countEntries = getFilesFromDirectory(cacheKeyMap, seekKeyInDB,
+              prefixPath, countEntries, numEntries, prefixKeyInDB);
+      // Seek the given key in dir table.
+      Table dirTable = metadataManager.getDirectoryTable();
+      TableIterator<String, ? extends Table.KeyValue<String, OmDirectoryInfo>>
+              iterator = dirTable.iterator();
+
+      iterator.seek(seekKeyInDB);
+
+      while (iterator.hasNext() && numEntries - countEntries > 0) {
+        String entryInDb = iterator.key();
+        OmDirectoryInfo dirInfo = iterator.value().getValue();
+        if (!isImmediateChild(dirInfo.getParentObjectID(), prefixKeyInDB)) {
+          break;
+        }
+
+        if (recursive) {
+          // for recursive list all the entries
+          prefixPath = OMFileRequest.getAbsolutePath(prefixPath,
+                  dirInfo.getName());
+          OmKeyInfo omKeyInfo = OMFileRequest.getOmKeyInfo(volumeName,
+                  bucketName, dirInfo, prefixPath);
+          cacheKeyMap.put(entryInDb,
+                  new OzoneFileStatus(omKeyInfo, 0, true));
+          ++countEntries;
+          // files from this directory
+          seekKeyInDB = dirInfo.getObjectID() + OZONE_URI_DELIMITER;
+          countEntries = getFilesFromDirectory(cacheKeyMap, seekKeyInDB,
+                  prefixPath, countEntries, numEntries, prefixKeyInDB);
+        } else {
+          String dirName = OMFileRequest.getAbsolutePath(prefixPath,
+                  dirInfo.getName());
+          OmKeyInfo omKeyInfo = OMFileRequest.getOmKeyInfo(volumeName,
+                  bucketName, dirInfo, dirName);
+          cacheKeyMap.put(entryInDb,
+                  new OzoneFileStatus(omKeyInfo, 0, true));
+          countEntries++;
+        }
+        // move to next entry in the table
+        iterator.next();
+      }
+    } finally {
+      metadataManager.getLock().releaseReadLock(BUCKET_LOCK, volumeName,
+              bucketName);
+    }
+
+    int countEntries = 0;
+    // Convert results in cacheKeyMap to List
+    for (Map.Entry<String, OzoneFileStatus> entry : cacheKeyMap.entrySet()) {
+      // No need to check if a key is deleted or not here, this is handled
+      // when adding entries to cacheKeyMap from DB.
+      if (args.getRefreshPipeline()) {
```

Comment:
We also don't need args.getRefreshPipeline() here due to HDDS-3658 change.

Reply:
Done